### PR TITLE
Grateful tock

### PR DESCRIPTION
**Pull request by jduss4**
## Description

A first stab at creating an API endpoint for "grateful tock," discussed in https://github.com/18F/tock/issues/824

- adds a field for `timecard.submitted_date` in order to determine if a timecard was submitted on time or late
- creates `api/submissions/NUM.json` endpoint to count punctual timecards by user
  - all counts are reported back for those timecards which were on time, not just those users with a 100% score

## Additional information

Several questions should be addressed with this change:

- should we automatically populate the `submitted_time` for timecards prior to this change?
- what happens if someone is on vacation / out sick and does not submit before the reporting period ends?

Furthermore, we should consider what behavior we want here vs in Charlie. For example, right now if you requested "5" reporting periods ago, you might get some users with "4" punctual timecards and others with "3" or fewer -- do we want to display the "4" or only display those with 100% timeliness?  Would we prefer that this was filtered on Charlie's side for reporting?

### CODE REVIEW

**Comment by jduss4** [comment]
NOTE:  I am not a Python nor a Django developer and likely there are syntax / style / query construction faux pas that should be altered in these commits, so please feel free to bring the critiques

**Comment by Jkrzy** [comment]
Thanks @jduss4! are you still working on this or is it ready for 😻 ? I see it's still marked as a draft.

**Comment by jduss4** [comment]
Hi @Jkrzy I suppose I left it in draft because I wasn't clear how well what I have done so far fit the requirements.  Specifically I don't know what to do about these two things:

- should we automatically populate the submitted_time for timecards prior to this change?
- what happens if someone is on vacation / out sick and does not submit before the reporting period ends?

Depending on the answers to those things, I may need to make some alterations.  Otherwise the code is probably reviewable now and should also be functional if you want to test anything out!  If testing manually, you will need to go into the admin interface for timecards and set some of the `submitted_date` fields to be during the respective reporting periods.

**Comment by Jkrzy** [comment]
Let's make an issue for our friend [Charlie](https://github.com/18F/charlie/issues) to consume this new endpoint and work some magic in Slack.

**Comment by Jkrzy** [comment]
> Should we automatically populate the submitted_time for timecards prior to this change?

In practice timecards are not commonly modified after submission so setting the value to `Timecard.modified` would could approximate submission date time but I don't think it's necessary. Let's only track this going forward to have the data be the most reliable.

 > What happens if someone is on vacation / out sick and does not submit before the reporting period ends?

We don't have any plans to track `why` a timecard was submitted before/after the end of a reporting period. Unless they Tocked in advance or let the folks in #tock know to tock on their behalf. That's OK.

File: tock/hours/models.py
Comment (jduss4) [inline tock/hours/models.py:232]: I just realized that I think I need to subtract from the `number_of_periods` because of how the Django limit / offset logic works.

`[:5]` => 5th result
`[5:10]` => *6th* through 10th result

So if the user puts in "5 weeks" I think this will need to be `[4:5]` to catch the 5th result.

**Comment by jduss4** [comment]
Now accepting feedback about if you want me to squash some commits :)

File: tock/api/urls.py
Comment (Jkrzy) [inline tock/api/urls.py:18]: Opportunity here to borrow the approach being used just above this for the ProjectInstanceView and avoid needing regular expressions. We'll also be able to skip casting the inbound value to an integer ourselves in the view.
```suggestion
    path(
        'submissions/<int:num_past_reporting_periods>.json',
```

File: tock/api/views.py
Comment (Jkrzy) [inline tock/api/views.py:197]: Grabbing the earliest reporting period may be a bit simpler here if we used negative indexing to grab the `last` object in the list instead of modifying `get_most_recent_periods`. This works by first evaluating the QuerySet to a plain old `list` so we can use negative indexing. 


```suggestion
        earliest_reporting_period = list(ReportingPeriod.get_most_recent_periods(
            number_of_periods=rp_num,
        ))[-1]
```

File: tock/api/views.py
Comment (Jkrzy) [inline tock/api/views.py:327]: Love this approach!!

File: tock/api/views.py
Comment (Jkrzy) [inline tock/api/views.py:192]: A bit of boundary handling would be a great next step here, as-is this throws an exception if someone attempts to ask "How many timecards, since X timecards ago have been submitted on time" for values of X greater than we have reporting periods.

File: tock/hours/models.py
Comment (Jkrzy) [inline tock/hours/models.py:233]: Noted above but this will break for any input where the `starting_index` is larger than the number of reporting periods in the system.

**Review by Jkrzy — CHANGES_REQUESTED** [review]
:amaze: @jduss4 Thanks so much for putting this together, really appreciate the docs and test coverage.

I left a few comments throughout where I saw potential opportunities to simplify things a bit. Would love your feedback on those. Overall this looks great!

File: api-docs/README.md
Comment (jduss4) [reply api-docs/README.md:34]: "and a count of timecards submitted on time by past reporting period?"

Comment (Jkrzy) [reply api-docs/README.md:34]: I was thinking it would be helpful for folks to know what Tock itself is considering as `on-time` or `punctual` here. Even just a plain-language description of this implementation would be helpful as I don't think there's an existing and consistently shared understanding of a date/time by which a timecard must be submitted.

File: tock/api/urls.py
Comment (jduss4) [reply tock/api/urls.py:18]: This is way better, thanks!

File: api-docs/README.md
Comment (jduss4) [reply api-docs/README.md]: I added a bit to clarify when a timecard is considered "on time" as added in this PR (if my assumptions are wrong then we should change some code!)

File: tock/api/views.py
Comment (jduss4) [reply tock/api/views.py]: This is definitely simpler. I went with this suggestion and removed the logic I added to the reporting period model for `number_of_results`

File: tock/api/views.py
Comment (jduss4) [reply tock/api/views.py]: Since switching to the `list()[-1]` suggestion, I believe there should no longer be any problems with bigger-than-reality requests

File: tock/hours/models.py
Comment (jduss4) [reply tock/hours/models.py]: I believe this is resolved as well

**Comment by jduss4** [comment]
I have one question before this gets merged:  what is the difference between a user and user_data?  I wasn't clear on whether I should be using `user.id` or `user_data` for the JSON response.

https://github.com/18F/tock/pull/1255/files#diff-a8868cbdadbfa18685d8ea42750df6024f25dc7ef80b5421320cd208d0759361R73

**Comment by Jkrzy** [comment]
Great question, and the answer is :django-pony: related! 

Within Tock, `user` refers to [django's standard `User` model](https://docs.djangoproject.com/en/2.2/ref/contrib/auth/#user-model)

`user_data` refers to `tock.employees.models.UserData`, a model defined within Tock that holds "extra" attributes we want to associate with the `User` model. It's a `OneToOne` relationship in Tock and most commonly accessed with syntax like  -- To get an employee's start date -- `user.user_data.start_date`.

In this specific issue, `user.id` is almost certainly fine until we recognize otherwise, though renaming the field to `user_id` may make it a bit more clear.

**Comment by jduss4** [comment]
I see!  Thanks, that's good to know.  I can change the field if you think that would be best.  I believe I was imitating `user` from the `TimecardSerializer` but have no particular attachment to it

**Review by Jkrzy — APPROVED** [review]
Thanks @jduss4!

**Pull request merged**
